File: core/src/main/java/y2k/joyreactor/presenters/MessagesPresenter.java
package y2k.joyreactor.presenters;

import y2k.joyreactor.Message;
import y2k.joyreactor.services.MessageService;
import y2k.joyreactor.services.requests.SendMessageRequest;

import java.util.List;

/**
 * Created by y2k on 01/10/15.
 */
public class MessagesPresenter extends Presenter {

    private View view;
    private MessageService service;

    public MessagesPresenter(View view, MessageService service) {
        this.view = view;
        this.service = service;

        // FIXME:
        reloadMessages(getUsername());

        getMessages().add(
                m -> reloadMessages(m.thread.userName),
                MessageThreadsPresenter.ThreadSelectedMessage.class);
    }

    public void reply(String message) {
        view.setIsBusy(true);
        new SendMessageRequest(getUsername())
                .request(message)
                .subscribe(s -> reloadMessages(getUsername()), Throwable::printStackTrace);
    }

    private void reloadMessages(String username) {
        view.setIsBusy(true);
        service.queryAsync(username)
                .subscribe(messages -> {
                    view.updateMessages(messages);
                    view.setIsBusy(false);
                }, Throwable::printStackTrace);
    }

    private String getUsername() {
        return "user500"; // TODO:
    }

    public interface View {

        void updateMessages(List<Message> messages);

        void setIsBusy(boolean isBusy);
    }
}

File: core/src/main/kotlin/y2k/joyreactor/presenters/MessagesPresenter.kt
package y2k.joyreactor.presenters

import y2k.joyreactor.model.Message
import y2k.joyreactor.common.subscribeOnMain
import y2k.joyreactor.services.BroadcastService
import y2k.joyreactor.services.LifeCycleService
import y2k.joyreactor.services.UserMessagesService
import y2k.joyreactor.services.requests.SendMessageRequest

/**
 * Created by y2k on 01/10/15.
 */
class MessagesPresenter(
    private val view: MessagesPresenter.View,
    private val service: UserMessagesService,
    private val lifeCycleService: LifeCycleService) {

    private var currentUsername: String? = null

    init {
        lifeCycleService.add(BroadcastService.ThreadSelectedMessage::class) {
            currentUsername = it.thread.userName
            reloadMessages()
        }
    }

    fun reply(message: String) {
        view.setBusy(true)
        view.clearMessage()
        currentUsername?.let {
            SendMessageRequest(it)
                .request(message)
                .subscribeOnMain(
                    { reloadMessages() },
                    { view.setBusy(false) })
        }
    }

    private fun reloadMessages() {
        currentUsername?.let {
            view.setBusy(true)
            service
                .getMessages(it)
                .subscribeOnMain({
                    view.updateMessages(it)
                    view.setBusy(false)
                }, {
                    view.setBusy(false)
                })
        }
    }

    interface View {

        fun updateMessages(messages: List<Message>)

        fun setBusy(isBusy: Boolean)

        fun clearMessage()
    }
}